File: Turquoise/UI/VC/TQGroupSelectionTableViewCell.m
#import "TQGroupSelectionTableViewCell.h"

#import "TQUserInfoManager.h"

@class TQNNTPGroup;

@implementation TQGroupSelectionTableViewCell

+ (UIColor *)evenColor {
  return [UIColor colorWithRed:8. / 255.
                         green:20. / 255.
                          blue:50. / 255.
                         alpha:1];
}

+ (UIColor *)oddColor {
  return [UIColor colorWithRed:8. / 255.
                         green:20. / 255.
                          blue:0. / 255.
                         alpha:1];
}

- (void)awakeFromNib {
  [super awakeFromNib];

  [self updateSubscriptionStatus];

  // TODO: make this notification system smarter.
  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(userSubscriptionsDidUpdate:)
                                               name:kUserSubscriptionsDidChangeNotification
                                             object:nil];
}

- (void)updateSubscriptionStatus {
  BOOL subscribed = NO;
  if (_group) {
    subscribed = [[TQUserInfoManager sharedInstance] isSubscribedToGroup:_group];
  }
  _statusIconView.image = subscribed ? [UIImage imageNamed:@"check1.png"]
                                     : nil;
}

- (void)userSubscriptionsDidUpdate:(NSNotification *)notification {
  [self updateSubscriptionStatus];
}

- (void)setGroup:(TQNNTPGroup *)group {
  _group = group;

  [self updateSubscriptionStatus];
  if (!_group) {
    _groupNameLabel.attributedText = nil;
    return;
  }

  NSInteger numArticles = ABS(_group.maxArticleNo - _group.minArticleNo) + 1;
  NSString *numArticlesString = [NSString stringWithFormat:@"(%ld)", numArticles];
  NSString *moderatedString = nil;
  if (_group.moderated) {
    moderatedString = @" [moderated]";
  }

  NSString *text = [NSString stringWithFormat:@"%@ %@%@",
                    _group.groupId,
                    numArticlesString,
                    (moderatedString ?: @"")];

  NSMutableAttributedString *attributedText = [[NSMutableAttributedString alloc] initWithString:text];
  [attributedText addAttribute:NSForegroundColorAttributeName
                         value:[UIColor colorWithRed:0 green:.5 blue:0 alpha:1]
                         range:[text rangeOfString:numArticlesString]];
  if (moderatedString) {
    [attributedText addAttribute:NSForegroundColorAttributeName
                           value:[UIColor redColor]
                           range:[text rangeOfString:moderatedString]];
  }

  _groupNameLabel.attributedText = attributedText;
}

@end

File: Turquoise/UI/VC/TQGroupSelectionTableViewCell.swift
import Foundation
import UIKit

class TQGroupSelectionTableViewCell : UITableViewCell {
  var group: TQNNTPGroup? {
    didSet {
      self.updateSubscriptionStatus()

      guard let group = self.group else {
        self.groupNameLabel.attributedText = nil
        return
      }

      let numArticles = abs(group.maxArticleNo - group.minArticleNo) + 1
      let numArticlesString = String(numArticles)
      var moderatedString: String?
      if group.moderated {
        moderatedString = " [moderated]"
      }

      let text = "\(group.groupId) \(numArticlesString)\(moderatedString ?? "")"

      let attributedText = NSMutableAttributedString(string: text)
        attributedText.addAttribute(NSAttributedString.Key.foregroundColor,
                                  value: UIColor(displayP3Red: 0, green: 0.5, blue: 0, alpha: 1),
                                  range: (text as NSString).range(of: numArticlesString))
      if let moderatedString = moderatedString {
        attributedText.addAttribute(NSAttributedString.Key.foregroundColor,
                                    value: UIColor.red,
                                    range: (text as NSString).range(of: moderatedString))
      }

      self.groupNameLabel.attributedText = attributedText
    }
  }

  @IBOutlet var statusIconView: UIImageView!
  @IBOutlet var groupNameLabel: UILabel!

  class func evenColor() -> UIColor {
    return UIColor(displayP3Red: 8.0 / 255.0,
                   green: 20.0 / 255.0,
                   blue: 50.0 / 255.0,
                   alpha: 1)
  }

  class func oddColor() -> UIColor {
    return UIColor(displayP3Red: 8.0 / 255.0,
                   green: 20.0 / 255.0,
                   blue: 0.0 / 255.0,
                   alpha: 1)
  }


  override func awakeFromNib() {
    super.awakeFromNib()

    self.updateSubscriptionStatus()

    // TODO: make this notification system smarter.
    NotificationCenter.default.addObserver(self,
                                           selector: #selector(userSubscriptionsDidUpdate(_:)),
                                           name: TQUserInfoManager.sharedInstance.userSubscriptionsDidChangeNotification,
                                           object: nil)
  }

  func updateSubscriptionStatus() {
    var subscribed = false
    if let group = self.group {
      subscribed = TQUserInfoManager.sharedInstance.isSubscribedTo(group: group)
    }
    self.statusIconView.image = subscribed ? UIImage(named: "check1.png")
                                           : nil
  }





    @objc func userSubscriptionsDidUpdate(_ notification: Notification) {
    self.updateSubscriptionStatus()
  }


}
